#### Imbalanced Dataset
A dataset with an unequal class distribution

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('credit_data.csv')

In [4]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [9]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [11]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

Here the dataset is clearly imbalanced as 284315 vs 492

0 --> Legit transactions
1 --> Fraudulent transactions

In [15]:
# Separating the legit and frauduelent transactions
legit = df[df.Class == 0]
fraud = df[df.Class == 1]

In [16]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


##### Undersampling
To handle imbalanced data - where equal no. of datapoints are taken from both classes

In [18]:
# Build a sample dataset containing similar distribution of Legit & Fraudulent
legit_sample = legit.sample(n=492)

In [19]:
print(legit_sample.shape)

(492, 31)


In [24]:
new_df = pd.concat([legit_sample, fraud], axis=0)
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
233590  147622.0  1.971258 -1.188438 -1.514333 -0.864551 -0.602179 -0.845563   
238869  149860.0 -4.028415  2.499500 -0.342332 -0.584266 -1.165208  1.042512   
168524  119268.0  1.342173 -0.990355 -1.362156 -1.785239  1.399576  3.944058   
5401      5356.0 -4.334220  3.819447 -2.209402  0.239475 -2.678888 -1.255879   
153897  100550.0 -0.010989 -0.109187  1.008510 -2.006929 -0.160389 -0.045351   

              V7        V8        V9  ...       V21       V22       V23  \
233590 -0.300239 -0.203904 -0.269803  ... -0.004474 -0.438612  0.240780   
238869 -1.568259  1.545433  0.327984  ...  0.705829 -0.471782  0.200557   
168524 -1.092216  0.640449 -2.022517  ... -0.076306  0.471845  0.297959   
5401   -2.026311  3.176342  0.098502  ...  0.225706 -0.078912  0.343918   
153897 -0.126517  0.020184  0.665073  ... -0.093760  0.569351  0.045162   

             V24       V25       V26       V27       V28  Amount  Class  
233590  0.616406 -0.301030 -0.462243 -0.070268 -0.031896  139.59      0  
238869 -1.083718  0.386518 -0.337136 -0.455923 -0.204197   14.43      0  
168524  0.708649 -1.401920  0.077675  0.060019 -0.146190   25.00      0  
5401    0.465950 -0.085603  0.186568 -0.892170 -0.179815   32.81      0  
153897 -0.522902 -1.322695  0.301605  0.214989  0.249819   19.99      0  

[5 rows x 31 columns]

In [27]:
new_df.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [31]:
new_df['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64